### DISCUSSION TITLE
Ownership in Supabase — how to automatically set 'owner_id' upon new inserts?

### DISCUSSION THREAD

**Discussion by adamhp** [supabase/supabase#4368] [Questions] [answered]
Is there anyway to set user ownership, (e.g.) `owner_id`, when someone creates an object, so that the user does not have to submit their own `id` along with each object?

I have a pretty straight forward use case, of a typical CRUD app with user ownership. Users can create, read, update, and delete arbitrary objects... let's just call it the `objects` table. The `objects` table has `owner_id` which is a FK to `user.id`, and several other text columns that are just data.

At the moment, from my React frontend, I'm sending along the user id with the request, so that the RLS policy can check it and restrict operations on only 'owned' objects by that user.

Is there a better way to handle this concept?

Ideally, the RLS policy would just require authenticated users on inserts for `objects` table, but some other function would also automatically get `uid()` (the user's id) and apply it as `owner_id`, so that all subsequent RLS (read, update, delete) could check authentication _and_ `uid() == owner_id`. Maybe there is a way to accomplish this with triggers or functions or something?

**Comment by GaryAustin1** [comment, marked as answer]
Normally a user token is passed with calls to the database and that token sets the user in the database for that call so you can use things like auth.uid() in your RLS.  

Then you can also set a default value for your owner_id column to uid() and not pass it in on your insert/updates.  If you did pass it in, it also would have to match the current user thru auth.uid() in the RLS for operation to work.

**Reply by adamhp** [reply]
Thank you. That default value trick is exactly what I needed.

**Comment by timmy-newtron** [comment]
Sorry for reopening this. I stumpled upon the same problem (setting the owner_id column to the authenticated users uuid).
setting the default value to 'uid()' does not work. Supabase web UI tells me that "function uid() does not exist".
When exchaning uid() to auth.uid() the error does not show up, however the owner_id column stays empty when creating data as an authenticated user.

**Reply by GaryAustin1** [reply]
Are you inserting null or a value for it?   The insert will override the default.

**Reply by timmy-newtron** [reply]
Thank you for your quick reply. No, I am not inserting anything in the column.

**Reply by timmy-newtron** [reply]
Extra piece of information: i have RLS disabled for that table. Could that have an impact?

**Reply by GaryAustin1** [reply]
Well you will probably have to show your code.  The Supabase Discord is usually a better place to get more interactive help than here.
Not sure if you have default auth.uid(), how it would get changed unless your insert has that column as something else.

**Reply by timmy-newtron** [reply]
I guess I found my mistake. I wasnt making supabase requests as an authenticated user in the first place. I use supabase in my own backend and pass through the user token that i receive from the frontend. I was reusing my supabase client in the backend (singleton) and wasnt creating one per user request.

**Reply by fendyjong** [reply]
Its "auth.uid()" for the default value.

https://stackoverflow.com/questions/69995024/postgrest-add-authenticated-user-id-on-insert
![Screen Shot 2023-08-03 at 9 45 18 AM](https://github.com/supabase/supabase/assets/30851388/8378bdd3-1b78-49c8-9e00-4dfb8ebced31)

**Reply by ammar188** [reply]
For anyone who comes here in the future, don't forget to set your value to not allow nulls so that whenever you receive nothing it is replaced by default value and then do not send null or anything else and it should work just fine.

**Reply by o1Suleyman** [reply]
> For anyone who comes here in the future, don't forget to set your value to not allow nulls so that whenever you receive nothing it is replaced by default value and then do not send null or anything else and it should work just fine.

doesn't work for me, it only works if I pass auth.uid() as the default value

**Reply by o1Suleyman** [reply]
> Its "auth.uid()" for the default value.
> 
> https://stackoverflow.com/questions/69995024/postgrest-add-authenticated-user-id-on-insert ![Screen Shot 2023-08-03 at 9 45 18 AM](https://private-user-images.githubusercontent.com/30851388/257986318-8378bdd3-1b78-49c8-9e00-4dfb8ebced31.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NjUxMzYsIm5iZiI6MTczMjQ2NDgzNiwicGF0aCI6Ii8zMDg1MTM4OC8yNTc5ODYzMTgtODM3OGJkZDMtMWI3OC00OWM4LTllMDAtNGRmYjhlYmNlZDMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI0VDE2MTM1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEzNDk2MTk5NmFhN2YwNzg1N2JkNzUwMDMzZmQyMDZmNzE1NDc0ZWZkNzgwNjliNjkwNzk2NzRkYWMzYzVmN2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NGvxoyNMh0fJsdAmQVpU8oXX_5puqGzTck2YGoP_PU8)

thanks, but why is this not in the documentation or mentioned anywhere?
